rust/otap-dataflow/crates/otap/src/retry_processor.rs

```diff
@@ -290,52 +228,123 @@ impl Processor<OtapPdata> for RetryProcessor {
+                match data.push_reply_to(
+                    Interest::NackOnly,
+                    effect_handler.processor_id().index(),
+                    ContextData::new(0),
+                ) {
+                    Ok(_) => {
+                        // Reply_to pushed
+                        effect_handler.send_message(data).await?;
+                        Ok(())
+                    }
+                    Err(err) => {
+                        // Reply_to not pushed, e.g., deadline already exceeded.
+                        effect_handler
+                            .reply_nack(
+                                data.return_node_id(),
+                                NackMsg::new(data, err.to_string(), false, None),
+                            )
+                            .await?;
+                        Ok(())
+                    }
                 }
-                Ok(())
             }
             Message::Control(control_msg) => match control_msg {
-                NodeControlMsg::Ack { id } => {
-                    self.acknowledge(id);
-                    Ok(())
+                NodeControlMsg::Ack(_) => {
+                    // This component does not subscribe to Acks, they pass directly
+                    // to the next interested component.
+                    unreachable!();
                 }
-                NodeControlMsg::Nack { id, reason, pdata } => {
-                    self.handle_nack(id, reason, pdata, effect_handler).await
+                NodeControlMsg::Nack(mut nack) => {
+                    // Backward path: NACK processing - PEEK at retry state to decide action
+                    let reply = nack.refused.take_reply();
+                    let frame = reply
+                        .ok_or_else(|| effect_handler.internal_error("missing reply state"))?;
+                    let mut rstate: RetryState = frame.data.into();
+
+                    // If the error is permanent or too many retries.
+                    // If the payload is empty: the effect is also
+                    // permanent, as by intentionaly failing to return
+                    // the data.
+                    let has_failed = if nack.permanent {
+                        nack.reason = format!("cannot retry permanent: {}", nack.reason);
+                        true
+                    } else if rstate.retries >= self.config.max_retries {
+                        nack.reason = format!("max retries reached: {}", nack.reason);
+                        true
+                    } else if nack.refused.is_empty() {
+                        nack.reason = format!("retry internal error: {}", nack.reason);
+                        nack.permanent = true;
+                        nack.code = Some(Code::Internal as i32);
+                        true
+                    } else {
+                        false
+                    };
+                    let return_node_id = nack.refused.return_node_id();
```

Comment:
IMO, knowing where to send the nack message next should not be the responsibility of the node.